app/models/piece.rb

```diff
@@ -86,6 +79,15 @@ def move_obstructed?(to_row, to_col)
     false
   end
 
+  def self_check?(to_row, to_col)
+    from_row = row
+    from_col = col
+    update(row: to_row, col: to_col)
```

Comment:
Since this actually commits the data to the database, it's no longer "atomic", so there's a possibility that the database could get into an inconsistent state if the process fails between the two `update` calls, either from something external like a crash or shutdown, or from an error on line 86.  Ideally, this would be refactored to avoid committing anything to the DB